### DISCUSSION TITLE
Injecting additional fields into the response.

### DISCUSSION THREAD

**Discussion by deepakguruu** [graphql-java/graphql-java#2383] [Q&A] [answered]
Is it possible to inject additional fields into the response?
I have a requirement where I need to inject some data into the response when a directive is encountered.
Can I do this directly from the DirectiveWiring or do I need to wire up some instrumentation to do the same?

**Comment by bbakerman** [comment, marked as answer]
Where do you need the extra data....as part of a fields data or on the request as a whole??


Datafetchers do per field data....instrumentation can change the over all result


If you outline more on you use case this will help

**Reply by deepakguruu** [reply]
My use case was to add a custom implementation of the deprecated tag, so that warnings will be thrown in the response(overall request not part of fields).
My current implementation is as follows, a deprecated directive implementation which adds the deprecated field to a global list maintained in the context object and an instrumentation which instruments the execution result which appends the deprecated fields list to the extensions part of the result.
Any suggestions to improve this would be welcome @bbakerman

**Reply by bbakerman** [reply]
This is pretty much the pattern - you need to share a common place (context object) that your fetchers can access and then ina  last step (instrument response) you need to add that data in.

Typically this is put into the graphql `extensions` object - this is the spec place for extra data to a query that is not asked for direct (goes in `data`) or errors (goes in `errors`)

We have thought about adding a way to "write to a common "extensions" place before but its tricky because its a map  and how to merge entries is hard to do well as a library.

So your approach is spot on

**Reply by bbakerman** [reply]
For example here is a grahql reply from our Atlassian gateway and its got extenstion data for certain things

```
{
  "data": {
    "tenantContexts": [
      {
        "cloudId": "xxxxxxxxxxxxxx"
      }
    ]
  },
  "extensions": {
    "resultComplexity": {
      "totalNodeCount": 0,
      "serviceNodeCounts": {
        "tcs": 0
      },
      "fieldRenamesCount": 0,
      "typeRenamesCount": 0
    },
    "gateway": {
      "request_id": "638faea2e8ad964",
      "schema": {
        "schemaName": "central",
        "commitHash": "bbf66560aca7a47fe84b8ebed694370762d30776",
        "schemaHash": "",
        "snapshot": "572",
        "buildNumber": "4270",
        "buildTimestamp": "2021-06-21T05:24:45.670554Z"
      },
      "micros_env": "stg-west",
      "invokedServices": [
        "tcs"
      ]
    }
  }
}

**Reply by deepakguruu** [reply]
Alright!
Thanks for the response!!
